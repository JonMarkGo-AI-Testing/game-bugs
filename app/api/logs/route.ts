import { NextResponse } from 'next/server';
import * as db from '../../lib/db';

export async function GET() {
  try {
    // Get all logs
    const logs = db.getAllLogs();
    console.log('Fetched logs:', logs); // Debug log

    // Parse JSON strings back to objects
    const parsedLogs = logs.map(log => ({
      ...log,
      details: log.details ? JSON.parse(log.details) : null,
      gameState: log.game_state ? JSON.parse(log.game_state) : null,
      // Ensure timestamp is properly formatted
      timestamp: new Date(log.timestamp),
    }));

    return NextResponse.json(parsedLogs);
  } catch (error) {
    console.error('Error fetching logs:', error);
    return NextResponse.json(
      { error: 'Error fetching logs', details: error.message }, 
      { status: 500 }
    );
  }
}

export async function POST(request: Request) {
  try {
    const log = await request.json();
    console.log('Received log data:', log); // Debug log

    // Validate required fields
    if (!log.id || !log.type || !log.message || !log.severity || !log.category) {
      console.error('Missing required fields:', { log });
      return NextResponse.json(
        { error: 'Missing required fields' },
        { status: 400 }
      );
    }

    // Insert the log
    const result = db.insertLog(
      log.id,
      new Date(log.timestamp).toISOString(),
      log.type,
      log.message,
      log.severity,
      log.category,
      JSON.stringify(log.details || null),
      log.stack || null,
      log.userAgent || null,
      JSON.stringify(log.gameState || null)
    );
    
    console.log('Log inserted successfully:', result); // Debug log
    return NextResponse.json({ success: true, id: log.id });
  } catch (error) {
    console.error('Error creating log:', error);
    return NextResponse.json(
      { 
        error: 'Error creating log', 
        details: error.message,
        stack: error.stack 
      }, 
      { status: 500 }
    );
  }
}

export async function DELETE() {
  try {
    db.clearAllLogs();
    return NextResponse.json({ success: true });
  } catch (error) {
    console.error('Error clearing logs:', error);
    return NextResponse.json(
      { error: 'Error clearing logs', details: error.message }, 
      { status: 500 }
    );
  }
}