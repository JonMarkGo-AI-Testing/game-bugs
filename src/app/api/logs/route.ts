import { NextResponse } from 'next/server';
import { insertLog, getAllLogs, getLogsByType, getLogsBySeverity, clearAllLogs, getLogStats } from '@/lib/db';

export async function POST(request: Request) {
  try {
    const log = await request.json();
    console.log('Received log data:', log); // Debug log

    // Validate required fields
    if (!log.id || !log.type || !log.message || !log.severity || !log.category) {
      console.error('Missing required fields:', { log });
      return NextResponse.json(
        { error: 'Missing required fields in log data' },
        { status: 400 }
      );
    }

    // Safely prepare game state
    let gameStateStr = null;
    if (log.gameState) {
      try {
        gameStateStr = JSON.stringify(log.gameState);
      } catch (e) {
        console.error('Error stringifying game state:', e);
        return NextResponse.json(
          { error: 'Invalid game state data' },
          { status: 400 }
        );
      }
    }

    const result = insertLog.run(
      log.id,
      new Date(log.timestamp).toISOString(),
      log.type,
      log.message,
      log.severity,
      log.category,
      JSON.stringify(log.details || null),
      log.stack || null,
      log.userAgent || null,
      gameStateStr
    );
    
    return NextResponse.json({ success: true });
  } catch (error) {
    console.error('Error creating log:', error);
    const errorMessage = error instanceof Error ? error.message : 'Unknown error';
    return NextResponse.json(
      { error: 'Error creating log', details: errorMessage }, 
      { status: 500 }
    );
  }
}

export async function GET() {
  try {
    // Get all logs
    const logs = getAllLogs.all();

    // Parse JSON strings back to objects
    const parsedLogs = logs.map(log => ({
      ...log,
      details: log.details ? JSON.parse(log.details) : null,
      gameState: log.game_state ? JSON.parse(log.game_state) : null,
      // Ensure timestamp is properly formatted
      timestamp: new Date(log.timestamp),
    }));

    return NextResponse.json(parsedLogs);
  } catch (error) {
    console.error('Error fetching logs:', error);
    return NextResponse.json({ error: 'Error fetching logs', details: error.message }, { status: 500 });
  }
}

export async function DELETE() {
  try {
    clearAllLogs.run();
    return NextResponse.json({ success: true });
  } catch (error) {
    console.error('Error clearing logs:', error);
    return NextResponse.json(
      { error: 'Error clearing logs' }, 
      { status: 500 }
    );
  }
}