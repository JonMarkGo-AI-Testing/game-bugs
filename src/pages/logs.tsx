import { LogDashboard } from '../components/LogDashboard';
import Layout from '../components/Layout';

export default function LogsPage() {
  return (
    <Layout>
      <LogDashboard />
    </Layout>
  );
}