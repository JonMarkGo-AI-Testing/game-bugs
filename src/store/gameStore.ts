import { create } from 'zustand';

export type LogSeverity = 'low' | 'medium' | 'high';

export type LogEntry = {
  id: string;
  timestamp: Date;
  type: 'error' | 'info' | 'warning';
  message: string;
  severity: LogSeverity;
  category: string;
  details?: any;
  stack?: string;
  userAgent?: string;
  gameState?: {
    score: number;
    snakeLength: number;
    position: { x: number; y: number };
  };
};

type GameStore = {
  score: number;
  highScore: number;
  logs: LogEntry[];
  loading: boolean;
  error: string | null;
  addLog: (entry: Omit<LogEntry, 'id' | 'timestamp'>) => Promise<void>;
  updateScore: (newScore: number) => void;
  clearLogs: () => Promise<void>;
  fetchLogs: () => Promise<void>;
  getLogStats: () => {
    total: number;
    byType: Record<LogEntry['type'], number>;
    bySeverity: Record<LogSeverity, number>;
  };
};

export const useGameStore = create<GameStore>((set, get) => ({
  score: 0,
  highScore: 0,
  logs: [],
  loading: false,
  error: null,
  
  addLog: async (entry) => {
    const newLog = {
      id: Math.random().toString(36).substring(7),
      timestamp: new Date(),
      userAgent: typeof navigator !== 'undefined' ? navigator.userAgent : '',
      ...entry,
    };

    try {
      // First, try to persist to SQLite
      const response = await fetch('/api/logs', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(newLog),
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || 'Failed to save log');
      }

      // If successful, update Zustand state
      set((state) => ({
        logs: [newLog, ...state.logs],
        error: null
      }));
    } catch (error) {
      console.error('Failed to persist log:', error);
      set({ error: 'Failed to save log to database' });
      throw error; // Re-throw to let the component handle the error
    }
  },

  fetchLogs: async () => {
    set({ loading: true, error: null });
    try {
      console.log('Fetching logs from API...');
      const response = await fetch('/api/logs');
      
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || 'Failed to fetch logs');
      }
      
      const logs = await response.json();
      console.log('Received logs:', logs);
      
      // Ensure timestamps are Date objects
      const parsedLogs = logs.map((log: any) => ({
        ...log,
        timestamp: new Date(log.timestamp),
        details: typeof log.details === 'string' ? JSON.parse(log.details) : log.details,
        gameState: typeof log.game_state === 'string' ? JSON.parse(log.game_state) : log.gameState,
      }));
      
      console.log('Parsed logs:', parsedLogs);
      set({ logs: parsedLogs, loading: false, error: null });
    } catch (error) {
      console.error('Failed to fetch logs:', error);
      set({ 
        error: error instanceof Error ? error.message : 'Failed to fetch logs',
        loading: false,
        logs: [] // Clear logs on error to avoid stale data
      });
    }
  },

  updateScore: (newScore) =>
    set((state) => ({
      score: newScore,
      highScore: Math.max(state.highScore, newScore),
    })),

  clearLogs: async () => {
    set({ loading: true, error: null });
    try {
      const response = await fetch('/api/logs', { 
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
        },
      });
      
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.error || 'Failed to clear logs');
      }
      
      set({ logs: [], loading: false, error: null });
    } catch (error) {
      console.error('Failed to clear logs:', error);
      set({ 
        error: error instanceof Error ? error.message : 'Failed to clear logs',
        loading: false 
      });
      throw error;
    }
  },

  getLogStats: () => {
    const logs = get().logs;
    return {
      total: logs.length,
      byType: logs.reduce((acc, log) => {
        acc[log.type] = (acc[log.type] || 0) + 1;
        return acc;
      }, {} as Record<LogEntry['type'], number>),
      bySeverity: logs.reduce((acc, log) => {
        acc[log.severity] = (acc[log.severity] || 0) + 1;
        return acc;
      }, {} as Record<LogSeverity, number>),
    };
  },
}));