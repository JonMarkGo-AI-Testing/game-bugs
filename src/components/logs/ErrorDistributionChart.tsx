import React from 'react';
import {
  ResponsiveContainer,
  PieChart,
  Pie,
  Cell,
  Tooltip,
  Legend
} from 'recharts';
import { LogEntry } from '../../store/gameStore';
import { prepareErrorDistribution } from '../../utils/logUtils';

interface ErrorDistributionChartProps {
  logs: LogEntry[];
}

const COLORS = [
  '#8B5CF6', // violet
  '#EC4899', // pink
  '#14B8A6', // teal
  '#10B981', // emerald
  '#6366F1', // indigo
  '#84CC16', // lime
  '#A855F7', // purple
  '#06B6D4', // cyan
  '#D946EF', // fuchsia
  '#34D399', // emerald-400
];

export const ErrorDistributionChart: React.FC<ErrorDistributionChartProps> = ({ logs }) => {
  return (
    <div className="bg-white p-4 rounded-lg shadow">
      <h2 className="text-lg font-semibold mb-4">Top Error Categories</h2>
      <div className="h-[300px]">
        <ResponsiveContainer width="100%" height="100%">
          <PieChart>
            <Pie
              data={prepareErrorDistribution(logs)}
              dataKey="count"
              nameKey="category"
              cx="50%"
              cy="50%"
              outerRadius={100}
              label={({ name, value, percent }) => `${name} (${(percent * 100).toFixed(0)}%)`}
            >
              {prepareErrorDistribution(logs).map((entry, index) => (
                <Cell 
                  key={`cell-${index}`} 
                  fill={COLORS[index % COLORS.length]} 
                />
              ))}
            </Pie>
            <Tooltip 
              formatter={(value, name) => [`${value} errors`, name]}
            />
            <Legend 
              layout="vertical" 
              align="right"
              verticalAlign="middle"
            />
          </PieChart>
        </ResponsiveContainer>
      </div>
    </div>
  );
}; 