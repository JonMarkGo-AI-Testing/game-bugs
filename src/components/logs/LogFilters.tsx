import React from 'react';
import { LogEntry, LogSeverity } from '../../store/gameStore';

interface LogFiltersProps {
  filter: LogEntry['type'] | 'all';
  setFilter: (filter: LogEntry['type'] | 'all') => void;
  severityFilter: LogSeverity | 'all';
  setSeverityFilter: (severity: LogSeverity | 'all') => void;
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

export const LogFilters: React.FC<LogFiltersProps> = ({
  filter,
  setFilter,
  severityFilter,
  setSeverityFilter,
  searchTerm,
  setSearchTerm,
}) => {
  return (
    <div className="bg-white rounded-lg shadow mb-6">
      <div className="p-4 border-b">
        <div className="grid grid-cols-1 md:grid-cols-3 gap-4">
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">Type</label>
            <div className="flex space-x-2">
              {['all', 'error', 'warning', 'info'].map((type) => (
                <button
                  key={type}
                  onClick={() => setFilter(type as any)}
                  className={`px-3 py-1 rounded ${
                    filter === type
                      ? 'bg-gray-200 text-gray-800'
                      : 'text-gray-600 hover:bg-gray-100'
                  }`}
                >
                  {type.charAt(0).toUpperCase() + type.slice(1)}
                </button>
              ))}
            </div>
          </div>
          
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">Severity</label>
            <div className="flex space-x-2">
              {['all', 'low', 'medium', 'high'].map((severity) => (
                <button
                  key={severity}
                  onClick={() => setSeverityFilter(severity as any)}
                  className={`px-3 py-1 rounded ${
                    severityFilter === severity
                      ? 'bg-gray-200 text-gray-800'
                      : 'text-gray-600 hover:bg-gray-100'
                  }`}
                >
                  {severity.charAt(0).toUpperCase() + severity.slice(1)}
                </button>
              ))}
            </div>
          </div>

          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">Search</label>
            <input
              type="text"
              value={searchTerm}
              onChange={(e) => setSearchTerm(e.target.value)}
              placeholder="Search logs..."
              className="w-full px-3 py-1 border rounded focus:outline-none focus:ring-2 focus:ring-blue-500"
            />
          </div>
        </div>
      </div>
    </div>
  );
}; 