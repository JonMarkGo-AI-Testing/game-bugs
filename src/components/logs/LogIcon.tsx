import React from 'react';
import { AlertTriangle, Info, XCircle } from 'lucide-react';
import { LogEntry } from '../../store/gameStore';

export const LogIcon: React.FC<{ type: LogEntry['type'] }> = ({ type }) => {
  switch (type) {
    case 'error':
      return <XCircle className="w-5 h-5 text-red-500" />;
    case 'warning':
      return <AlertTriangle className="w-5 h-5 text-yellow-500" />;
    case 'info':
    default:
      return <Info className="w-5 h-5 text-blue-500" />;
  }
}; 