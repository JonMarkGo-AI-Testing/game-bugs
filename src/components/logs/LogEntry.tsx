import React from 'react';
import { Clock, Tag } from 'lucide-react';
import { LogEntry as LogEntryType } from '../../store/gameStore';
import { LogIcon } from './LogIcon';
import { SeverityBadge } from './SeverityBadge';

export const LogEntryComponent: React.FC<{ log: LogEntryType }> = ({ log }) => {
  return (
    <div className="p-4 bg-white rounded-lg shadow border border-gray-200">
      <div className="flex items-start gap-4">
        <LogIcon type={log.type} />
        <div className="flex-1">
          <div className="flex items-center gap-2 mb-2">
            <h3 className="font-medium">{log.message}</h3>
            <SeverityBadge severity={log.severity} />
            <span className="text-sm text-gray-500">
              <Tag className="w-4 h-4 inline mr-1" />
              {log.category}
            </span>
          </div>
          <div className="flex items-center text-gray-500 text-sm mb-2">
            <Clock className="w-4 h-4 mr-1" />
            {new Date(log.timestamp).toLocaleString()}
          </div>
          {log.gameState && (
            <div className="mb-2 text-sm text-gray-600">
              Game State: Score: {log.gameState.score}, Length: {log.gameState.snakeLength}
            </div>
          )}
          {log.details && (
            <pre className="mt-2 p-2 bg-gray-50 rounded text-sm overflow-x-auto">
              {JSON.stringify(log.details, null, 2)}
            </pre>
          )}
          {log.stack && (
            <div className="mt-2 p-2 bg-red-50 rounded text-sm overflow-x-auto font-mono">
              {log.stack}
            </div>
          )}
        </div>
      </div>
    </div>
  );
}; 