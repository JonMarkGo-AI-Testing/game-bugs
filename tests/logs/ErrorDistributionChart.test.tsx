import React from 'react';
import { render, screen } from '@testing-library/react';
import { ErrorDistributionChart } from '../../src/components/logs/ErrorDistributionChart';
import { LogEntry } from '../../src/store/gameStore';

// Mock Recharts components
jest.mock('recharts', () => ({
  ResponsiveContainer: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  PieChart: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  Pie: ({ data, label }: { data: any[], label: Function }) => (
    <div data-testid="pie">
      {data.map((entry, index) => (
        <div key={index} data-testid="pie-label">
          {/* This tests the label format in the pie chart (the percentage labels) */}
          {label({ 
            name: entry.category, 
            value: entry.count, 
            percent: entry.count / data.reduce((acc, curr) => acc + curr.count, 0) 
          })}
        </div>
      ))}
    </div>
  ),
  Cell: () => null,
  Tooltip: ({ formatter }: { formatter: Function }) => {
    // Test the tooltip format
    const [value, name] = formatter(2, 'collision');
    return (
      <div data-testid="tooltip">
        {/* This should render as "collision: 2 errors" */}
        {name}: {value}
      </div>
    );
  },
  Legend: () => (
    <div data-testid="legend">
      {/* Legend should just show the category name */}
      collision
    </div>
  ),
}));

describe('ErrorDistributionChart', () => {
  describe('snapshot tests', () => {
    it('matches snapshot with single category', () => {
      const logs = createMockLog('collision', 2);
      const { container } = render(<ErrorDistributionChart logs={logs} />);
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with multiple categories', () => {
      const logs = [
        ...createMockLog('collision', 2),
        ...createMockLog('network', 2)
      ];
      const { container } = render(<ErrorDistributionChart logs={logs} />);
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with empty logs', () => {
      const { container } = render(<ErrorDistributionChart logs={[]} />);
      expect(container).toMatchSnapshot();
    });
  });

  const createMockLog = (category: string, count: number = 1): LogEntry[] => 
    Array.from({ length: count }, () => ({
      id: Math.random().toString(),
      timestamp: Date.now(),
      type: 'error',
      severity: 'high',
      message: 'Test error',
      category
    }));

  it('formats tooltip and legend correctly for collision errors', () => {
    const logs: LogEntry[] = [
      ...createMockLog('collision', 2), // 2 collision errors
    ];

    render(<ErrorDistributionChart logs={logs} />);

    // Check tooltip format ("collision: 2 errors")
    const tooltip = screen.getByTestId('tooltip');
    expect(tooltip).toHaveTextContent('collision: 2 errors');
    expect(tooltip).not.toHaveTextContent('NaN');

    // Check legend format (just "collision")
    const legend = screen.getByTestId('legend');
    expect(legend).toHaveTextContent('collision');
    expect(legend).not.toHaveTextContent('(');
    expect(legend).not.toHaveTextContent(')');
    expect(legend).not.toHaveTextContent('%');

    // Log for debugging
    console.log('Tooltip content:', tooltip.textContent);
    console.log('Legend content:', legend.textContent);
  });

  it('formats pie chart labels correctly', () => {
    const logs: LogEntry[] = [
      ...createMockLog('collision', 2),
    ];

    render(<ErrorDistributionChart logs={logs} />);

    const pieLabels = screen.getAllByTestId('pie-label');
    
    // Check pie label format ("collision (100%)")
    expect(pieLabels[0]).toHaveTextContent('collision (100%)');
    expect(pieLabels[0]).not.toHaveTextContent('NaN');

    // Log for debugging
    console.log('Pie label content:', pieLabels[0].textContent);
  });

  it('handles multiple error categories', () => {
    const logs: LogEntry[] = [
      ...createMockLog('collision', 2),
      ...createMockLog('network', 2),
    ];

    render(<ErrorDistributionChart logs={logs} />);

    const tooltip = screen.getByTestId('tooltip');
    const legend = screen.getByTestId('legend');

    // Check tooltip still shows correct count for individual category
    expect(tooltip).toHaveTextContent('collision: 2 errors');

    // Check legend shows just the category name
    expect(legend).toHaveTextContent('collision');
    expect(legend).not.toHaveTextContent('%');

    // Log for debugging
    console.log('Multiple categories - Tooltip:', tooltip.textContent);
    console.log('Multiple categories - Legend:', legend.textContent);
  });

  it('handles empty logs array', () => {
    render(<ErrorDistributionChart logs={[]} />);
    
    const pie = screen.getByTestId('pie');
    expect(pie).toBeInTheDocument();
    
    const pieLabels = screen.queryAllByTestId('pie-label');
    expect(pieLabels).toHaveLength(0);
  });
});
