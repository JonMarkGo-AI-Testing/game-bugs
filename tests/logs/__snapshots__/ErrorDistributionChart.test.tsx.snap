// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorDistributionChart snapshot tests matches snapshot with empty logs 1`] = `
<div>
  <div
    class="bg-white p-4 rounded-lg shadow"
  >
    <h2
      class="text-lg font-semibold mb-4"
    >
      Top Error Categories
    </h2>
    <div
      class="h-[300px]"
    >
      <div>
        <div>
          <div
            data-testid="pie"
          />
          <div
            data-testid="tooltip"
          >
            collision
            : 
            2 errors
          </div>
          <div
            data-testid="legend"
          >
            collision
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorDistributionChart snapshot tests matches snapshot with multiple categories 1`] = `
<div>
  <div
    class="bg-white p-4 rounded-lg shadow"
  >
    <h2
      class="text-lg font-semibold mb-4"
    >
      Top Error Categories
    </h2>
    <div
      class="h-[300px]"
    >
      <div>
        <div>
          <div
            data-testid="pie"
          >
            <div
              data-testid="pie-label"
            >
              collision (50%)
            </div>
            <div
              data-testid="pie-label"
            >
              network (50%)
            </div>
          </div>
          <div
            data-testid="tooltip"
          >
            collision
            : 
            2 errors
          </div>
          <div
            data-testid="legend"
          >
            collision
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorDistributionChart snapshot tests matches snapshot with single category 1`] = `
<div>
  <div
    class="bg-white p-4 rounded-lg shadow"
  >
    <h2
      class="text-lg font-semibold mb-4"
    >
      Top Error Categories
    </h2>
    <div
      class="h-[300px]"
    >
      <div>
        <div>
          <div
            data-testid="pie"
          >
            <div
              data-testid="pie-label"
            >
              collision (100%)
            </div>
          </div>
          <div
            data-testid="tooltip"
          >
            collision
            : 
            2 errors
          </div>
          <div
            data-testid="legend"
          >
            collision
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
